File: week07/18/prob4.c
#include <stdio.h>
#include <stdlib.h>

int canDivide(int A[], int N, int K, int max_block_sum) {
    int blocks_used = 1;  // We need at least one block
    int current_block_sum = 0;

    for (int i = 0; i < N; i++) {
        // If adding A[i] to the current block exceeds the max_block_sum, create a new block
        if (current_block_sum + A[i] > max_block_sum) {
            blocks_used++;
            current_block_sum = A[i];  // Start new block with A[i]
            if (blocks_used > K) {
                return 0;  // More than K blocks needed
            }
        } else {
            current_block_sum += A[i];  // Add to the current block
        }
    }
    return 1;  // It's possible to split the array into K or fewer blocks
}

// Main solution function
int solution(int K, int M, int A[], int N) {
    int max_element = 0;
    int total_sum = 0;

    // Calculate the maximum element and the sum of the array
    for (int i = 0; i < N; i++) {
        if (A[i] > max_element) {
            max_element = A[i];
        }
        total_sum += A[i];
    }

    // Binary search for the minimal large sum
    int low = max_element;
    int high = total_sum;
    int result = total_sum;

    while (low <= high) {
        int mid = (low + high) / 2;
        // Check if we can divide the array with `mid` as the max large sum
        if (canDivide(A, N, K, mid)) {
            result = mid;  // This is a valid solution, try for a smaller one
            high = mid - 1;
        } else {
            low = mid + 1;  // This is too small, try a larger value
        }
    }

    return result;
}

File: week07/23/prob4.py
def can_divide(A, K, max_block_sum):
    block_count = 1
    current_block_sum = 0
    
    for num in A:
        if current_block_sum + num > max_block_sum:
            # Start a new block
            block_count += 1
            current_block_sum = num
            if block_count > K:
                return False
        else:
            current_block_sum += num
    
    return True

def minimize_large_sum(K, M, A):
    # Binary search boundaries
    left = max(A)  # Minimum possible large sum is the max element
    right = sum(A)  # Maximum possible large sum is the sum of the array
    
    while left <= right:
        mid = (left + right) // 2
        if can_divide(A, K, mid):
            right = mid - 1  # Try a smaller large sum
        else:
            left = mid + 1  # Increase the large sum
    
    return left

# Example usage
K = 3
M = 5
A = [2, 1, 5, 1, 2, 2, 2]
result = minimize_large_sum(K, M, A)
print("Minimal large sum:", result)
